README.md
# supermarket-pricing

This repository contains a Clojure based model for a supermarket pricing system. The aim is to represent a shopping basket and receipt in a way that allows for different pricing structures like:

* Three tins of beans for the price of two
* Onions for 29p / kg
* Two cans of coca-cola for £1
* Any 3 ales from the set {...} for £6

## Process

I started by considering the most basic system where the model can find the basket's total by summing the quantities of each item found in the basket times by the price of that item found in a prices database. The pricing system was steadily built up by including specific units for each item and the price per unit.

In order to include pricing offers, like what a supermarket might offer, the individual offer types were added in turn. The model is set up so that each offer has a different name, item and parameter values. To create a new offer the user just has to then write a new function to calculate the saving associated with that offer. Each offer follows the same format:

1. Find the appropriate items for that offer
2. Calculate the original price (as if the offer didn't exist)
3. Calculate the reduced price including the :offer-type offer
4. Find the difference

## Usage

To use the model a basket must be defined as a map:
```python
(def basket {:beans 3 :coke 3 :onions 0.2 :ale1 1 :ale2 2 :ale3 2 :ale4 1})
```

Then any offers the model uses can be defined using the offer functions:
```python
(def my-offer (offer-fixed-price "coke" 2 1))
```

To form the receipt use `form-receipt` with the basket and any offers:
```python
(def receipt (form-receipt basket my-offer))
```

To view the processed receipt with associated prices and savings:
```python
(def process-receipt receipt)
```

## Limitations and Improvements

There is no automatic checking of schema in this model. If a user were to use a basket that wasn't a map it would throw a non-descript error. To improve `prismatic-schema` could be used in order to define schemas for the input basket and prices database.

Currently the savings for each individual offer are not stored, only the sum of all the offers. To improve this, the offers in the receipt could have their respective sacvings associated with them inside the processed receipt.

For the `set-for-fixed-price` offer, the savings is calculated from the price of the first one with assumption all the items in the set are the same price. If the set has items of a different price then the offer function will need to be changed to find the minimum saving within the set.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-03-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-03-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/supermarket-pricing/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/supermarket-pricing/compare/0.1.0...0.1.1


test/supermarket_pricing/core_test.clj
(ns supermarket-pricing.core-test
  (:require [clojure.test :refer :all]
            [supermarket-pricing.core :refer :all]))

(def test-basket
  {:beans  4
   :coke   2
   :onions 0.7})

(def test-receipt-empty
  {:items [] :offers []})

(def test-receipt-no-offers
  {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}]
   :offers []})

(def test-receipt-fixed-price-offer
  {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}]
   :offers [{:type "fixed-price" :item "coke" :quantity 2 :price 1}]})

(def test-receipt-x-for-y-offer
  {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}]
   :offers [{:type "x-for-y" :item "beans" :x 3 :y 2}]})

(def test-receipt-multiple-offers
  {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}]
   :offers [{:type "fixed-price" :item "coke" :quantity 2 :price 1}
            {:type "x-for-y" :item "beans" :x 3 :y 2}]})

(def test-receipt-all-offers
  {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}
            {:name "ale1"   :quantity 1   :unit "bottle" :price-per-unit 3}
            {:name "ale2"   :quantity 2   :unit "bottle" :price-per-unit 3}
            {:name "ale3"   :quantity 2   :unit "bottle" :price-per-unit 3}
            {:name "ale4"   :quantity 1   :unit "bottle" :price-per-unit 3}]
   :offers [{:type "fixed-price" :item "coke" :quantity 2 :price 1}
            {:type "x-for-y" :item "beans" :x 3 :y 2}
            {:type "set-for-fixed-price" :item  #{"ale1" "ale2" "ale3"} :quantity 3 :price 6}]})

(deftest receipt-is-formed-correctly
  (testing "Form empty receipt"
    (is (= test-receipt-empty (form-receipt {}))))
  (testing "Form receipt with no offers"
    (is (= test-receipt-no-offers (form-receipt test-basket))))
  (testing "Form receipt with fixed-price offer"
    (is (= test-receipt-fixed-price-offer (form-receipt test-basket (offer-fixed-price "coke" 2 1)))))
  (testing "Form receipt with x-for-y offer"
    (is (= test-receipt-x-for-y-offer (form-receipt test-basket (offer-x-for-y "beans" 3 2)))))
  (testing "Form receipt with fixed-price offer"
    (is (= test-receipt-multiple-offers (form-receipt test-basket (offer-fixed-price "coke" 2 1) (offer-x-for-y "beans" 3 2))))))


(deftest price-of-receipt-is-calculated-correctly
  (testing "Price of empty receipt"
    (is (= "£0.00" (price-of-receipt test-receipt-empty))))
  (testing "Price of receipt with no offers"
    (is (= "£4.79" (price-of-receipt test-receipt-no-offers))))
  (testing "Price of receipt with fixed-price offer"
    (is (= "£4.39" (price-of-receipt test-receipt-fixed-price-offer))))
  (testing "Price of receipt with x-for-y offer"
    (is (= "£4.29" (price-of-receipt test-receipt-x-for-y-offer))))
  (testing "Price of receipt with multiple offers"
    (is (= "£3.89" (price-of-receipt test-receipt-multiple-offers))))
  (testing "Price of receipt with all offers"
    (is (= "£18.89" (price-of-receipt test-receipt-all-offers)))))

(deftest receipt-is-processed-correctly
  (is (= {:items  [{:name "beans"  :quantity 4   :unit "tin" :price-per-unit 0.5}
                   {:name "coke"   :quantity 2   :unit "can" :price-per-unit 0.7}
                   {:name "onions" :quantity 0.7 :unit "kg"   :price-per-unit 1.99}
                   {:name "ale1"   :quantity 1   :unit "bottle" :price-per-unit 3}
                   {:name "ale2"   :quantity 2   :unit "bottle" :price-per-unit 3}
                   {:name "ale3"   :quantity 2   :unit "bottle" :price-per-unit 3}
                   {:name "ale4"   :quantity 1   :unit "bottle" :price-per-unit 3}]
          :offers [{:type "fixed-price" :item "coke" :quantity 2 :price 1}
                   {:type "x-for-y" :item "beans" :x 3 :y 2}
                   {:type "set-for-fixed-price" :item  #{"ale1" "ale2" "ale3"} :quantity 3 :price 6}]
          :prices {:sub-total "£22.79" :total-savings "£3.90" :total-to-pay "£18.89"}}
         (process-receipt test-receipt-all-offers))))


src/supermarket_pricing/core.clj
(ns supermarket-pricing.core
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

(defn round
  "Round to two decimal places for currency"
  [d]
  (let [factor (Math/pow 10 2)]
    (/ (Math/round (* d factor)) factor)))

(def prices
  "Mock up of a database containing units and price per units for different products"
  {:beans  {:unit "tin"
            :price-per-unit 0.50}
   :coke   {:unit "can"
            :price-per-unit 0.70}
   :onions {:unit "kg"
            :price-per-unit 1.99}
   :ale1   {:unit "bottle"
            :price-per-unit 3}
   :ale2   {:unit "bottle"
            :price-per-unit 3}
   :ale3   {:unit "bottle"
            :price-per-unit 3}
   :ale4   {:unit "bottle"
            :price-per-unit 3}})

(def eg-basket
  "Example of a basket represented in a map data structure"
  {:beans  3
   :coke   3
   :onions 0.2
   :ale1   1
   :ale2   2
   :ale3   2
   :ale4   1})

(def eg-receipt
  "Example of a receipt created by 'form-receipt' represented in a map with two keys:
  'items'  - the items in the basket
  'offers' - the offers associated with the receipt"
  {:items  [{:name "beans"  :quantity 4   :unit "item"   :price-per-unit 0.5}
            {:name "coke"   :quantity 2   :unit "item"   :price-per-unit 0.7}
            {:name "onions" :quantity 0.7 :unit "kg"     :price-per-unit 1.99}
            {:name "ale1"   :quantity 1   :unit "bottle" :price-per-unit 3}
            {:name "ale2"   :quantity 2   :unit "bottle" :price-per-unit 3}
            {:name "ale3"   :quantity 2   :unit "bottle" :price-per-unit 3}
            {:name "ale4"   :quantity 1   :unit "bottle" :price-per-unit 3}]
   :offers [{:type "x-for-y"             :item "beans"                 :x 3        :y 2}
            {:type "fixed-price"         :item "coke"                  :quantity 2 :price 1}
            {:type "set-for-fixed-price" :item #{"ale1" "ale2" "ale3"} :quantity 3 :price 6}]})

(defn form-receipt
  "Function to form a receipt made up of information from:
  - prices database
  - customer's basket
  - any offers presented"
  [basket & offers]
  {:items (mapv #(do {:name (name %)
                      :quantity (% basket)
                      :unit (:unit (% prices))
                      :price-per-unit (:price-per-unit (% prices))})
                (keys basket))
   :offers (vec offers)})

(defn offer-fixed-price
  "Representation of a fixed-price offer: 'Get :quantity :item for :price pounds.'"
  [item quantity price]
  {:type "fixed-price"
   :item item
   :quantity quantity
   :price price})

(defn offer-x-for-y
  "Representation of an x-for-y offer: ':x units of :item for the price of :y'"
  [item x y]
  {:type "x-for-y"
   :item item
   :x x
   :y y})

(defn offer-set-for-fixed-price
  "Representation of a set-for-fixed-price offer: ':quantity units in set :offer-set for the price of :price'"
  [item quantity price]
  {:type "set-for-fixed-price"
   :item item
   :quantity quantity
   :price price})

(defn price-of-item
  "Calculate price of an :item in the items section of receipt"
  [item]
  (* (:quantity item) (:price-per-unit item)))

(defn savings-fixed-price
  "Function to calculate the saving associated with a fixed-price offer:
  1. Find the appropriate order for that offer
  2. Calculate the original price (as if the offer didn't exist)
  3. Calculate the reduced price including the fixed-price offer
  4. Find the difference"
  [items {:keys [item quantity price]}]
  (let [order     (first (filter #(= item (:name %)) items))
        og-price  (round (* (:quantity order) (:price-per-unit order)))
        new-price (round (+ (* (:price-per-unit order) (mod (:quantity order) quantity))
                            (* price (int (/ (:quantity order) quantity)))))]
    (round (- og-price new-price))))

(defn savings-set-for-fixed-price
  "Function to calculate the saving associated with a fixed-price offer:
  1. Find the appropriate order for that offer
  2. Calculate the original price (as if the offer didn't exist)
  3. Calculate the reduced price including the fixed-price offer
  4. Find the difference"
  [items {:keys [item quantity price] :as offer}]

  (let [order     (filter #(item (:name %)) items)
        og-price  (apply + (map #(round (* (:quantity %) (:price-per-unit %))) order))
        set-quantity (apply + (map :quantity order))
        set-price    (:price-per-unit (first order))
        new-price (round (+ (* set-price (mod set-quantity quantity))
                            (* price (int (/ set-quantity quantity)))))]
    (round (- og-price new-price))))

(defn savings-x-for-y
  "Function to calculate the saving associated with a x-for-y offer:
  1. Find the appropriate order for that offer
  2. Calculate the original price (as if the offer didn't exist)
  3. Calculate the reduced price including the x-for-y offer
  4. Find the difference"
  [items {:keys [item x y]}]
  (let [order     (first (filter #(= item (:name %)) items))
        og-price  (round (* (:quantity order) (:price-per-unit order)))
        new-price (round (+ (* (:price-per-unit order) (mod (:quantity order) x))
                            (* (* y (:price-per-unit order)) (int (/ (:quantity order) x)))))]
    (round (- og-price new-price))))

(defn calculate-all-savings
  "Function to calculate the total of all the different savings associated with the offers in the receipt"
  [{:keys [items offers]}]
  (apply + (for [{:keys [type] :as offer} offers]
             (case type
               "set-for-fixed-price" (savings-set-for-fixed-price items offer)
               "fixed-price"         (savings-fixed-price items offer)
               "x-for-y"             (savings-x-for-y items offer)
               :else                 0))))

(defn price-of-receipt
  "Function to find the final receipt price: sub-total - savings"
  [receipt]
  (let [sub-total (->> (:items receipt)
                       (map price-of-item)
                       (apply +)
                       round)
        savings   (calculate-all-savings receipt)]
    (str "£" (format "%.2f" (- sub-total savings)))))

(defn process-receipt
  "Function to show the final receipt"
  [receipt]
  (let [sub-total (->> (:items receipt)
                       (map price-of-item)
                       (apply +)
                       round)
        savings   (calculate-all-savings receipt)
        total     (- sub-total savings)]
    (assoc receipt :prices {:sub-total (str "£" (format "%.2f" sub-total))
                            :total-savings (str "£" (format "%.2f" savings))
                            :total-to-pay (str "£" (format "%.2f" total))})))
